bindings/java/c/evmc-vm.c

```diff
@@ -0,0 +1,116 @@
+#include <stdlib.h>
+#include <assert.h>
+#include <stdint.h>
+
+#include "evmc/loader.h"
+#include "evmc/helpers.h"
+#include "evmc-vm.h"
+#include "host.h"
+
+JNIEXPORT jobject JNICALL Java_org_ethereum_evmc_EvmcVm_init(JNIEnv* jenv, jclass jcls, jstring jfilename)
+{
+    struct evmc_vm *evm;
+    jint rs = (*jenv)->GetJavaVM(jenv, &jvm);
+    assert (rs == JNI_OK);
+    // load the EVM
+    const char *filename = (*jenv)->GetStringUTFChars(jenv, jfilename, 0);
+    if(filename != NULL){
+        enum evmc_loader_error_code loader_error;
+        evm = evmc_load_and_create(filename, &loader_error);
+        if (evm == NULL || loader_error != EVMC_LOADER_SUCCESS) {
+            const char *error_msg = evmc_last_error_msg();
+            jclass jclazz = (*jenv)->FindClass(jenv, "java/lang/AssertionError");
+            (*jenv)->ThrowNew(jenv, jclazz, error_msg);
+        }
+        (*jenv)->ReleaseStringUTFChars(jenv, jfilename, filename);
+    } else {
+        jclass jclazz = (*jenv)->FindClass(jenv, "java/lang/AssertionError");
+        (*jenv)->ThrowNew(jenv, jclazz, "JNI Error: filename cannot be NULL. \n");
+    }
+     return (*jenv)->NewDirectByteBuffer(jenv, (void*) evm, sizeof(struct evmc_vm));
+}
+
+JNIEXPORT jint JNICALL Java_org_ethereum_evmc_EvmcVm_abi_1version(JNIEnv *jenv, jclass jcls){
+    return EVMC_ABI_VERSION;
+}
+
+JNIEXPORT jstring JNICALL Java_org_ethereum_evmc_EvmcVm_name(JNIEnv *jenv, jclass jcls, jobject jevm){
+    struct evmc_vm *evm = (struct evmc_vm *) (*jenv)->GetDirectBufferAddress(jenv, jevm);
+    const char *evm_name = evmc_vm_name(evm);
+    return (*jenv)->NewStringUTF(jenv, evm_name);
+}
+
+JNIEXPORT jstring JNICALL Java_org_ethereum_evmc_EvmcVm_version(JNIEnv *jenv, jclass jcls, jobject jevm){
+    struct evmc_vm *evm = (struct evmc_vm *) (*jenv)->GetDirectBufferAddress(jenv, jevm);
+    const char *evm_version = evmc_vm_version(evm);
+    return (*jenv)->NewStringUTF(jenv, evm_version);
+}
+
+JNIEXPORT void JNICALL Java_org_ethereum_evmc_EvmcVm_destroy(JNIEnv *jenv, jclass jcls, jobject jevm){
+    struct evmc_vm *evm = (struct evmc_vm *) (*jenv)->GetDirectBufferAddress(jenv, jevm);
+    evmc_destroy(evm);
+}
+
+JNIEXPORT jobject JNICALL Java_org_ethereum_evmc_EvmcVm_execute(JNIEnv *jenv, jclass jcls, jobject jevm, jint jcontext_index, jint jrev, jobject jmsg, jobject jcode, jint jsize){
+    jobject jresult;
+    struct evmc_result* result = malloc (sizeof (struct evmc_result));
```

Comment:
Why do you need this allocated and why can't it live on the stack on this function?

Reply:
My choices are:
- allocate result on the heap
- allocate result on the stack and mark it static
- make result global

Otherwise, this line

```c
jresult = (*jenv)->NewDirectByteBuffer(jenv, (void*) result, sizeof(struct evmc_result));
```

assigns NULL to jresult.  As a result, i opted for the heap; however, I am now using an assert to check for NULL and freeing the memory.  I am open to other suggestions.

Reply:
I'm not sure I understand. To me it seems that `NewDirectByteBuffer` allocates and memory and copies the input to it. Isn't that what's going on?

Reply:
My understanding is that it allocates a byte-buffer object that is a wrapper around native memory that can be accessed directly in Java.  That being said, it creates a problem for me because I need to properly free the memory from the heap and that would require another call to some function that frees the memory. It is less trouble if I simply allocate the memory in Java via `ByteBuffer.allocateDirect()` and let the java GC handle it.  So I have now changed the code to reflect this.

(from the docs 👇)
**NewDirectByteBuffer**
Native code that calls this function and returns the resulting byte-buffer object to Java-level code should ensure that the buffer refers to a valid region of memory that is accessible for reading and, if appropriate, writing. An attempt to access an invalid memory location from Java code will either return an arbitrary value, have no visible effect, or cause an unspecified exception to be thrown.

Reply:
Oh, my mistake then. I do prefer the approach of letting Java handling that memory entirely, simplifies the C code.

Reply:
ok good deal.  i agree.  glad u brought it up bc it forced me to reconsider what i was doing.